File: file-templates/templates/MVVM/TableList.xctemplate/UIViewObjective-C/___FILEBASENAME___ViewModel.m
//___FILEHEADER___

#import "___FILEBASENAME___.h"
#import "___VARIABLE_productName___CellModel.h"

@interface ___VARIABLE_productName:identifier___ViewModel ()
@property (nonatomic, strong) NSObject *model;
@property (nonatomic, strong) NSMutableArray <___VARIABLE_productName___CellModel *>*cellModels;
@end

@implementation ___VARIABLE_productName___ViewModel

- (instancetype)initWithModel:(NSObject *)model {
    self = [super init];
    if (self) {
        self.model = model;
        [self onInit];
    }
    return self;
}

- (void)onInit {
    NSMutableArray *cellModels = @[].mutableCopy;
    
    for (int i = 0; i < 10; i++) {
        ___VARIABLE_productName___CellModel *cellModel = [[___VARIABLE_productName___CellModel alloc] initWithModel:self];
        [cellModels addObject:cellModel];
    }
    
    [self.cellModels addObjectsFromArray:cellModels];
}

#pragma mark - ___VARIABLE_productName:identifier___ViewModelDatasource

- (NSInteger)numberOfSections {
    return 1;
}

- (NSInteger)numberOfRowsInSection:(NSInteger)section {
    return self.cellModels.count;
}

- (___VARIABLE_productName___CellModel *)itemForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.cellModels.count) {
        return self.cellModels[indexPath.row];
    }
    return nil;
}

- (CGFloat)heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    ___VARIABLE_productName___CellModel *cellModel = [self itemForRowAtIndexPath:indexPath];
    return cellModel.rowHeight;
}

- (void)onSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    ___VARIABLE_productName___CellModel *cellModel = [self itemForRowAtIndexPath:indexPath];
}

#pragma mark - getter/setter

- (NSMutableArray *)cellModels {
    if (!_cellModels) {
        _cellModels = @[].mutableCopy;
    }
    return _cellModels;
}

@end

File: file-templates/templates/MVVM/List.xctemplate/UIViewSwift/___FILEBASENAME___ViewModel.swift
//  ___FILEHEADER___
//  Summary: ___VARIABLE_Summary___

import Foundation

/// 数据源
protocol ___VARIABLE_productName:identifier___ViewModelDatasource {
    func numberOfSections() -> Int
    
    func numberOfRowsInSection(_ section: Int) -> Int
    
    func itemForRowAt(_ indexPath: IndexPath) -> Any?
    
    func heightForRowAt(_ indexPath: IndexPath) -> CGFloat
}

/// 代理
@objc protocol ___VARIABLE_productName:identifier___ViewModelDelegate: AnyObject {

    @objc optional func doSomething()
}

/// 列表视图模型
final class ___VARIABLE_productName:identifier___ViewModel {
    
    // MARK: - Inputs
    
    weak var delegate: ___VARIABLE_productName:identifier___ViewModelDelegate?
    
    /// cellModel构成的数组
    private var cellModels: [AnyObject] = [] // 示例代码
    
    // MARK: - Outputs

    // MARK: - Init

    init() {

    }
}

// MARK: - ___VARIABLE_productName:identifier___ViewModelDatasource

extension ___VARIABLE_productName:identifier___ViewModel: ___VARIABLE_productName:identifier___ViewModelDatasource {

  func numberOfSections() -> Int {
        return 1
    }
    
    func numberOfRowsInSection(_ section: Int) -> Int {
        return self.cellModels.count
    }
    
    func itemForRowAt(_ indexPath: IndexPath) -> Any? {
        if indexPath.row < self.cellModels.count {
            return self.cellModels[indexPath.row]
        }
        return nil
    }
    
    func heightForRowAt(_ indexPath: IndexPath) -> CGFloat {
        let item = itemForRowAt(indexPath)
//        return item.rowHeight
        return 44.0
    }
}